### PR TITLE
feat: enhance validations on adding a new ruleset

### PR DESCRIPTION
**Pull request by troypoulter**
Relates to: #2.

- Validates that a new `ruleset` being added doesn't overlap with an existing ruleset.

image

- Ensure `startTime` is before `endTime` and throw a validation error otherwise, learnt how to use `refine` on the Zod schema and for the best result, needed to split the Zod objects into two and then merge them together using `and` - checkout the code comment for link to GitHub issue explaining it more.

image

- Ensure `dayOfWeek` is one of the expected values.

image

- Properly parse `multipler` as either a `string` or `number` and `transform` as appropriate, solving a bug where we couldn't _nicely_ set a default value as a `number` as previously we set the expected value to be a `string` which we then transformed.
- Moved content into their own files, like `interfaces/Ruleset` now contains the schema/interface code and `utils/constants` contains our re-usable values, so we can `map` over them more easily.

### CODE REVIEW

File: app/utils/constants.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
+import { Ruleset } from "../interfaces/Ruleset";
+
+export const TIME_OF_DAY_REGEX = new RegExp(/^([1-9]|1[0-2]):[0-5][0-9] (AM|PM)$/);
+export const DAYS_OF_WEEK = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"];
+export const TIME_OF_DAY_INCREMENTS = [
```
Comment (madrym) [inline app/utils/constants.ts:5]: Ah very nice :) was annoying having to scroll past all these each time haha, like this new layout

File: app/utils/utils.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    const exactMatch = rulesets.some((existingRuleset) => {
+      return (
+        newRuleset.dayOfWeek === existingRuleset.dayOfWeek &&
+        newRuleset.startTime === existingRuleset.startTime &&
+        newRuleset.endTime === existingRuleset.endTime
+      );
+    });
+  
+    if (exactMatch) {
+        return {isValid: false, errorMessage: 'The new ruleset exactly matches an existing ruleset.'};
```
Comment (madrym) [inline app/utils/utils.ts:32]: Nice rule here, no duplicates :)

**Review by madrym — APPROVED** [review]

File: app/utils/constants.ts (hunk 1 above)
Comment (troypoulter) [reply app/utils/constants.ts:5]: Oh for sure keeps it more central and easier for re-use!

File: app/interfaces/Ruleset.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    message: "Day of the week must be a valid day.",
+    }),
+  multipler: z
+  .union([z.number(), z.string()])
+  .transform((value) => (typeof value === "string" ? Number(value) : value))
+})
+
+export const RulesetSchema = baseRulesetSchema.and(rulesetDateSchema)
+
+export type Ruleset = z.infer<typeof RulesetSchema>
```
Comment (lysng) [inline app/interfaces/Ruleset.ts:61]: oh nice I will look to extrapolate the interface out like this also

File: app/utils/constants.ts [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        endTime: "5:00 PM",
+        multipler: 2.0,
+      },
+      {
+        name: "Weekend Hours",
+        dayOfWeek: "Sunday",
+        startTime: "9:00 AM",
+        endTime: "5:00 PM",
+        multipler: 2.0,
+      }] as Ruleset[];
```
Comment (lysng) [inline app/utils/constants.ts:105]: sneaky little enterline here needs to be aded (settings I think) nothing major

File: app/interfaces/Ruleset.ts (hunk 3 above)
Comment (troypoulter) [reply app/interfaces/Ruleset.ts:61]: This is the recommended way of doing it with `zod`, create the schema then infer a type, so you don't need to double define!

**Pull request merged**